tests/bsim/bluetooth/host/l2cap/reassembly/peer/src/peer.c

```diff
@@ -0,0 +1,613 @@
+
+	disconn = net_buf_add(buf, sizeof(*disconn));
+	disconn->handle = sys_cpu_to_le16(handle);
+	disconn->reason = reason;
+
+	send_cmd(BT_HCI_OP_DISCONNECT, buf, NULL);
+
+	WAIT_FOR_FLAG_UNSET(is_connected);
+	LOG_INF("Disconnected");
+}
+
+NET_BUF_POOL_DEFINE(acl_tx_pool, 5, BT_L2CAP_BUF_SIZE(200), 8, NULL);
+
+struct net_buf *alloc_l2cap_pdu(void)
+{
+	struct net_buf *buf;
+	uint16_t reserve;
+
+	buf = net_buf_alloc(&acl_tx_pool, K_FOREVER);
+	TEST_ASSERT(buf, "failed ACL allocation");
+
+	reserve = sizeof(struct bt_l2cap_hdr);
+	reserve += sizeof(struct bt_hci_acl_hdr) + BT_BUF_RESERVE;
+
+	net_buf_reserve(buf, reserve);
+
+	return buf;
+}
+
+static int send_acl(struct net_buf *buf, uint8_t flags)
+{
+	struct bt_hci_acl_hdr *hdr;
+
+	hdr = net_buf_push(buf, sizeof(*hdr));
+	hdr->handle = sys_cpu_to_le16(bt_acl_handle_pack(conn_handle, flags));
+	hdr->len = sys_cpu_to_le16(buf->len - sizeof(*hdr));
+
+	bt_buf_set_type(buf, BT_BUF_ACL_OUT);
+
+	k_sem_take(&acl_pkts, K_FOREVER);
+
+	return bt_send(buf);
+}
+
+static void push_l2cap_pdu_header(struct net_buf *dst, uint16_t len, uint16_t cid)
+{
+	struct bt_l2cap_hdr *hdr;
+
+	hdr = net_buf_push(dst, sizeof(*hdr));
+	hdr->len = sys_cpu_to_le16(len);
+	hdr->cid = sys_cpu_to_le16(cid);
+}
+
+static void send_l2cap_packet(struct net_buf *buf, uint16_t cid)
+{
+	push_l2cap_pdu_header(buf, buf->len, cid);
+	send_acl(buf, BT_ACL_START_NO_FLUSH);
+}
+
+static void prepare_controller(void)
+{
+	/* Initialize controller */
+	struct net_buf *rsp;
+
+	send_cmd(BT_HCI_OP_RESET, NULL, NULL);
+	send_cmd(BT_HCI_OP_LE_READ_BUFFER_SIZE, NULL, &rsp);
+	le_read_buffer_size_complete(rsp);
+
+	set_event_mask(BT_HCI_OP_SET_EVENT_MASK);
+	set_event_mask(BT_HCI_OP_LE_SET_EVENT_MASK);
+	set_random_address();
+}
+
+static void init_tinyhost(void)
+{
+	bt_enable_raw(&rx_queue);
+
+	/* Start the RX thread */
+	k_thread_create(&rx_thread_data, rx_thread_stack,
+			K_THREAD_STACK_SIZEOF(rx_thread_stack), rx_thread,
+			NULL, NULL, NULL, K_PRIO_PREEMPT(0), 0, K_NO_WAIT);
+	k_thread_name_set(&rx_thread_data, "HCI RX");
+
+	k_thread_priority_set(k_current_get(), K_PRIO_PREEMPT(0));
+
+	prepare_controller();
+}
+
+static void gatt_notify(void)
+{
+	static uint8_t data[] = NOTIFICATION_PAYLOAD;
+	uint16_t handle = GATT_HANDLE;
+	struct net_buf *buf = alloc_l2cap_pdu();
+
+	net_buf_add_u8(buf, BT_ATT_OP_NOTIFY);
+	net_buf_add_le16(buf, handle);
+	net_buf_add_mem(buf, data, sizeof(data));
+
+	LOG_INF("Sending complete notification");
+	send_l2cap_packet(buf, BT_L2CAP_CID_ATT);
+}
+
+/* Send all but the last fragment of a notification */
+static void gatt_notify_without_last_fragment(void)
+{
+	static uint8_t data[] = NOTIFICATION_PAYLOAD;
+	uint16_t handle = GATT_HANDLE;
+	struct net_buf *att_packet = alloc_l2cap_pdu();
+
+	/* Prepare (G)ATT notification packet */
+	net_buf_add_u8(att_packet, BT_ATT_OP_NOTIFY);
+	net_buf_add_le16(att_packet, handle);
+	net_buf_add_mem(att_packet, data, sizeof(data));
+
+	size_t on_air_size = 5;
```

Comment:
Where does the `on_air` value 5 come from? 
I would prefer that NOTIFICATION_PAYLOAD is longer than 27 bytes to guarantee that the controller cannot send the data in a single packet as some implementations may choose to buffer payload until nothing more can fit.

Reply:
That makes sense!